File: l33tcode/possible-bipartition.py
from collections import defaultdict
from typing import List, Dict, Set, Optional


class Solution:
    def possibleBipartition(self, N: int, dislikes: List[List[int]]) -> bool:
        def make_graph(dislikes: List[List[int]]) -> Dict[int, List[int]]:
            graph = defaultdict(list)
            for left, right in dislikes:
                graph[right].append(left)
                graph[left].append(right)

            return graph

        graph = make_graph(dislikes)
        colors: List[Optional[bool]] = [None] * (N + 1)

        def dfs(pos: int, color: bool) -> bool:
            if colors[pos] is not None:
                return colors[pos] == color

            colors[pos] = color

            result = True
            for neigh_pos in graph[pos]:
                result = result and dfs(neigh_pos, not color)

            return result

        result = True

        for pos in range(1, N + 1):
            if colors[pos] is None:
                result = result and dfs(pos, True)

        return result

File: l33tcode/possible-bipartition.go
package main

func color(node int, adjList [][]int, colors []int, visited []bool) bool {
	visited[node] = true

	for _, adjNode := range adjList[node] {
		if colors[node] == 0 && colors[adjNode] == 0 {
			colors[node] = 1
			colors[adjNode] = 2
		}

		if colors[node] == 1 {
			if colors[adjNode] == 2 || colors[adjNode] == 0 {
				colors[adjNode] = 2

				if visited[adjNode] {
					continue
				}

				if !color(adjNode, adjList, colors, visited) {
					return false
				}
			} else {
				return false
			}
		}

		if colors[node] == 2 {
			if colors[adjNode] == 1 || colors[adjNode] == 0 {
				colors[adjNode] = 1

				if visited[adjNode] {
					continue
				}

				if !color(adjNode, adjList, colors, visited) {
					return false
				}
			} else {
				return false
			}
		}
	}
	return true
}

func possibleBipartition(n int, dislikes [][]int) bool {
	colors := make([]int, n)
	visited := make([]bool, n)

	adjList := make([][]int, n)

	for _, dislike := range dislikes {
		src, dst := dislike[0]-1, dislike[1]-1

		adjList[src] = append(adjList[src], dst)
		adjList[dst] = append(adjList[dst], src)
	}

	for node, _ := range make([]int, n) {
		if !color(node, adjList, colors, visited) {
			return false
		}
	}

	return true
}
